fix: differentiate legacy and segwit types for p2sh when using `Descriptor::scriptpubkey_type`

qlrd opened the pull request:
* fix #93;

Add a check for `p2sh-p2wsh`, `p2sh-p2wpkh` and raise an exception for
unknow types.


With this fix i was able to do a proper recognition of nested segwit multisig wallet when using this method to load a descriptor.

odudex commented:
The commit from this pull request has now been merged into Embit's develop branch.

jdlcdl commented:
Note: To include this in release-notes since this alters api/expected-return where existing projects may be expecting "p2sh"

odudex commented:
Another issue:
In order to use `owns()` method from `descriptor.py` and correctly detect change and self transfers in nested types, `script_type()` from `script.py` must be updated accordingly to cover detailed p2sh: p2sh-p2wpkh and p2sh-p2wsh

tadeubas commented:
How Specter DIY and SS does that if it is not implemented?

qlrd commented:
> In order to use owns() method from descriptor.py and correctly detect change and self transfers in nested types, script_type() from script.py must be updated accordingly to cover detailed p2sh: p2sh-p2wpkh and p2sh-p2wsh

I think that something like this could be used in `owns` (but didn't tested):

```diff
- if psbt_scope.script_pubkey.script_type() != self.scriptpubkey_type():
+ if psbt_scope.script_pubkey.script_type() not in self.scriptpubkey_type():
```

> I believe script_type() can't retrieve more information than it does now. So another approach should be used

yup, but during the sign procedure we can check, as a lib,  if there's `redeem_script` and `redeem_script.script_type()` to assert p2sh, p2sh-p2wpkh, p2sh-p2wsh (i effectively did that).

> How Specter DIY and SS does that if it is not implemented?

It's better to ask @kdmukai and @k9ert.

qlrd commented:
> > I think that something like this could be used in `owns` (but didn't tested):
> > ```diff
> > - if psbt_scope.script_pubkey.script_type() != self.scriptpubkey_type():
> > + if psbt_scope.script_pubkey.script_type() not in self.scriptpubkey_type():
> > ```
> 
> It seems to be a good idea!

Changed. Also added some `is_using_nums` and `is_pk` method to help in multisig and pure compreesed keys, as well some tests (need to think in more cases and add some taproot cases).

qlrd commented:
> > I think that something like this could be used in `owns` (but didn't tested):
> > ```diff
> > - if psbt_scope.script_pubkey.script_type() != self.scriptpubkey_type():
> > + if psbt_scope.script_pubkey.script_type() not in self.scriptpubkey_type():
> > ```
> 
> It seems to be a good idea!
> 
> > yup, but during the sign procedure we can check, as a lib, if there's `redeem_script` and `redeem_script.script_type()` to assert p2sh, p2sh-p2wpkh, p2sh-p2wsh (i effectively did that).
> 
> With the above idea it might not be necessary.

Also started with some signet PSBTs for testing `Descriptor.owns`
